### ISSUE TITLE
Subscribe disconnect

### ISSUE THREAD

**Issue by kleunen** [zeromq/azmq#98]
I have created a small subscriber application which uses the curve security. However the subscriber keeps disconnecting after a couple of minutes. After a couple of minutes I always the ZMQ_EVENT_DISCONNECTED event. Perhaps this is related to the curve security ? 

This is the example I have created:

```
#include <azmq/socket.hpp>
#include <boost/asio.hpp>
#include <array>

namespace asio = boost::asio;

asio::io_service ios;
boost::shared_ptr<azmq::sub_socket> subscriber;
boost::shared_ptr<azmq::socket> monitor;
uint16_t monitor_event;

boost::array<char, 4096> buffer;

void ZeroMQHandleMonitor(boost::system::error_code const& error, azmq::socket::more_result_type result)
{
    if((monitor_event & ZMQ_EVENT_CONNECTED) == ZMQ_EVENT_CONNECTED)
        std::cout << "ZeroMQ connection established" << std::endl;

    if(monitor_event & ZMQ_EVENT_CONNECT_DELAYED)
        std::cout << "ZeroMQ connect delayed" << std::endl;

    if(monitor_event & ZMQ_EVENT_CONNECT_RETRIED)
        std::cout << "ZeroMQ connect retried" << std::endl;

    if(monitor_event & ZMQ_EVENT_LISTENING)
        std::cout << "ZeroMQ server listening" << std::endl;

    if(monitor_event & ZMQ_EVENT_BIND_FAILED)
        std::cout << "ZeroMQ bind failed" << std::endl;

    if(monitor_event & ZMQ_EVENT_ACCEPTED)
        std::cout << "ZeroMQ connection accepted" << std::endl;

    if(monitor_event & ZMQ_EVENT_ACCEPT_FAILED)
        std::cout << "ZeroMQ accept failed" << std::endl;

    if(monitor_event & ZMQ_EVENT_CLOSED)
        std::cout << "ZeroMQ connection closed" << std::endl;

    if((monitor_event & ZMQ_EVENT_CLOSE_FAILED) == ZMQ_EVENT_CLOSE_FAILED)
        std::cout << "ZeroMQ close failed" << std::endl;

    if((monitor_event & ZMQ_EVENT_DISCONNECTED) == ZMQ_EVENT_DISCONNECTED)
        std::cout << "ZeroMQ connection lost" << std::endl;

    if(monitor_event & ZMQ_EVENT_MONITOR_STOPPED)
        std::cout << "ZeroMQ monitor stopped" << std::endl;

    monitor->flush();
    monitor->async_receive_more(boost::asio::buffer(&monitor_event, sizeof(monitor_event)), ZeroMQHandleMonitor);
}

void message_received(boost::system::error_code const& ec, azmq::socket::more_result_type result)
{
    std::cout << std::string(buffer.data(), result.first) << std::endl;	
    subscriber->async_receive_more(boost::asio::buffer(buffer), message_received); 
}

int main(int argc, char** argv) 
{
	subscriber = boost::make_shared<azmq::sub_socket>(ios);
    
	subscriber->set_option(azmq::socket::tcp_keepalive(1));
    subscriber->set_option(azmq::socket::subscribe(""));
	
	monitor = boost::make_shared<azmq::socket>(ios, true);
    *monitor = subscriber->monitor(ios, ZMQ_EVENT_ALL);
    monitor->async_receive_more(boost::asio::buffer(&monitor_event, sizeof(monitor_event)), ZeroMQHandleMonitor);
	
        char client_public_key [41];
        char client_secret_key [41];
        zmq_curve_keypair(client_public_key, client_secret_key);

        std::vector<uint8_t> server_key =  ... ;
							
        subscriber->set_option(azmq::socket::curve_server(false));
        subscriber->set_option(azmq::socket::curve_publickey(client_public_key, 40));
        subscriber->set_option(azmq::socket::curve_privatekey(client_secret_key, 40));
        subscriber->set_option(azmq::socket::curve_serverkey(server_key.data(), server_key.size()));        
	
	subscriber->connect("tcp://exampleserver.com:11000");
	subscriber->async_receive_more(boost::asio::buffer(buffer), message_received);

	ios.run();

	return 0;
}

```

**Comment by oliora** [comment]
@kleunen have you tried this scenario without curve security? Is it also disconnecting after a couple of minutes?

**Comment by kleunen** [comment]
Hello Andrey,


I have not tried this, I will try this.


Wouter

Op 9-8-2015 om 0:47 schreef Andrey Upadyshev:

**Comment by kleunen** [comment]
Hello Andrey,


I have tried to reproduce the problem, however I am unable to do so now. 
The connection stays connected. I already had doubts before the problem 
is really in azmq, because azmq merely wraps the zmq library. And before 
I had used the zmq library and had not experienced this problem.

One thing I did not is that the zmq library did not appear to reconnect 
the connection when it was lost. So perhaps there is still a condition 
in the zmq library where it fails to attempt to reconnect.

I am sorry, but now I am unable to reproduce the problem.

Kind regards,

Wouter


Op 9-8-2015 om 0:47 schreef Andrey Upadyshev:
